File: src/jfan/fan/sys/FanInt.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   2 Dec 05  Brian Frank  Creation
//   4 Oct 08  Brian Frank  Refactor long into Long/FanInt
//
package fan.sys;

import fanx.serial.*;

/**
 * FanInt defines the methods for sys::Int
 *   sys::Int   =>  long primitive
 *   sys::Int?  =>  java.lang.Long
 */
public final class FanInt
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Long fromStr(String s) { return fromStr(s, 10, true); }
  public static Long fromStr(String s, long radix) { return fromStr(s, radix, true); }
  public static Long fromStr(String s, long radix, boolean checked)
  {
    try
    {
      return Long.valueOf(s, (int)radix);
    }
    catch (NumberFormatException e)
    {
      if (!checked) return null;
      throw ParseErr.make("Int", s).val;
    }
  }

  public static long random() { return random(null); }
  public static long random(Range r)
  {
    long v = random.nextLong();
    if (r == null) return v;
    if (v < 0) v = -v;
    long start = r.start();
    long end   = r.end();
    if (r.inclusive()) ++end;
    return start + (v % (end-start));
  }
  static final java.util.Random random = new java.security.SecureRandom();

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public static boolean equals(long self, Object obj)
  {
    if (obj instanceof Long)
      return self == ((Long)obj).longValue();
    else
      return false;
  }

  public static long compare(long self, Object obj)
  {
    long that = (Long)obj;
    if (self < that) return -1; return self == that ? 0 : +1;
  }

  public static long hash(long self)
  {
    return self;
  }

  public static Type type(long self)
  {
    return Sys.IntType;
  }

//////////////////////////////////////////////////////////////////////////
// Operators
//////////////////////////////////////////////////////////////////////////

  public static long negate(long self)
  {
    return -self;
  }

  public static long inverse(long self)
  {
    return ~self;
  }

  public static long mult(long self, long x)
  {
    return self * x;
  }

  public static long div(long self, long x)
  {
    return self / x;
  }

  public static long mod(long self, long x)
  {
    return self % x;
  }

  public static long plus(long self, long x)
  {
    return self + x;
  }

  public static long minus(long self, long x)
  {
    return self - x;
  }

  public static long and(long self, long x)
  {
    return self & x;
  }

  public static long or(long self, long x)
  {
    return self | x;
  }

  public static long xor(long self, long x)
  {
    return self ^ x;
  }

  public static long lshift(long self, long x)
  {
    return self << x;
  }

  public static long rshift(long self, long x)
  {
    return self >> x;
  }

  public static long increment(long self)
  {
    return self+1;
  }

  public static long decrement(long self)
  {
    return self-1;
  }

//////////////////////////////////////////////////////////////////////////
// Math
//////////////////////////////////////////////////////////////////////////

  public static long abs(long self)
  {
    if (self >= 0) return self;
    return -self;
  }

  public static long min(long self, long that)
  {
    if (self <= that) return self;
    return that;
  }

  public static long max(long self, long that)
  {
    if (self >= that) return self;
    return that;
  }

  public static boolean isEven(long self)
  {
    return (self % 2) == 0;
  }

  public static boolean isOdd(long self)
  {
    return (self % 2) != 0;
  }

/////////////////////////////////////////////////////////////////////////
// Char
//////////////////////////////////////////////////////////////////////////


  public static boolean isSpace(long self)
  {
    try
    {
      return (self < 128 && (charMap[(int)self] & SPACE) != 0);
    }
    catch (ArrayIndexOutOfBoundsException e)
    {
      // should be very rare to use this method with negative
      // numbers, so don't take the hit every call
      return false;
    }
  }

  public static boolean isAlpha(long self)
  {
    try
    {
      return self < 128 && (charMap[(int)self] & ALPHA) != 0;
    }
    catch (ArrayIndexOutOfBoundsException e)
    {
      // should be very rare to use this method with negative
      // numbers, so don't take the hit every call
      return false;
    }
  }

  public static boolean isAlphaNum(long self)
  {
    try
    {
      return (self < 128 && (charMap[(int)self] & ALPHANUM) != 0);
    }
    catch (ArrayIndexOutOfBoundsException e)
    {
      // should be very rare to use this method with negative
      // numbers, so don't take the hit every call
      return false;
    }
  }

  public static boolean isUpper(long self)
  {
    return 'A' <= self && self <= 'Z';
  }

  public static boolean isLower(long self)
  {
    return 'a' <= self && self <= 'z';
  }

  public static long upper(long self)
  {
    if ('a' <= self && self <= 'z')
      return self & ~0x20L;
    else
      return self;
  }

  public static long lower(long self)
  {
    if ('A' <= self && self <= 'Z')
      return self | 0x20L;
    else
      return self;
  }

  public static boolean isDigit(long self)
  {
    return '0' <= self && self <= '9';
  }

  public static boolean isDigit(long self, long r)
  {
    if (self < 0 || self >= 128) return false;

    int val = (int)self;
    int radix = (int)r;
    if (radix == 10)
    {
      return ((charMap[val] & DIGIT) != 0);
    }

    if (radix == 16)
    {
      return ((charMap[val] & HEX) != 0);
    }

    if (radix <= 10)
    {
      return '0' <= val && val <= ('0'+radix);
    }
    else
    {
      if ((charMap[val] & DIGIT) != 0) return true;
      int x = val-10;
      if ('a' <= val && val <= 'a'+x) return true;
      if ('A' <= val && val <= 'A'+x) return true;
      return false;
    }
  }

  public static Long toDigit(long self)
  {
    if (0 <= self && self <= 9) return pos[(int)self+ '0'];
    return null;
  }

  public static Long toDigit(long self, long radix)
  {
    if (self < 0 || self >= radix) return null;

    if (self < 10) return pos[(int)self + '0'];
    return pos[(int)self - 10 + 'a'];
  }

  public static Long fromDigit(long self)
  {
    if ('0' <= self && self <= '9') return pos[(int)self - '0'];
    return null;
  }

  public static Long fromDigit(long self, long r)
  {
    if (self < 0 || self >= 128) return null;
    int val = (int)self;

    int radix = (int)r;
    int ten = radix < 10 ? radix : 10;
    if ('0' <= val && val < '0'+ten) return pos[val - '0'];
    if (radix > 10)
    {
      int alpha = radix-10;
      if ('a' <= val && val < 'a'+alpha) return pos[val + 10 - 'a'];
      if ('A' <= val && val < 'A'+alpha) return pos[val + 10 - 'A'];
    }
    return null;
  }

  public static boolean equalsIgnoreCase(long self, long ch)
  {
    return (self | 0x20L) == (ch | 0x20L);
  }

  static final byte[] charMap = new byte[128];
  static final int SPACE    = 0x01;
  static final int UPPER    = 0x02;
  static final int LOWER    = 0x04;
  static final int DIGIT    = 0x08;
  static final int HEX      = 0x10;
  static final int ALPHA    = UPPER | LOWER;
  static final int ALPHANUM = UPPER | LOWER | DIGIT;
  static
  {
    charMap[' ']  |= SPACE;
    charMap['\n'] |= SPACE;
    charMap['\r'] |= SPACE;
    charMap['\t'] |= SPACE;
    charMap['\f'] |= SPACE;

    // alpha characters
    for (int i='a'; i<='z'; ++i) charMap[i] |= LOWER;
    for (int i='A'; i<='Z'; ++i) charMap[i] |= UPPER;

    // digit characters
    for (int i='0'; i<='9'; ++i) charMap[i] |= DIGIT;

    // hex characters
    for (int i='0'; i<='9'; ++i) charMap[i] |= HEX;
    for (int i='a'; i<='f'; ++i) charMap[i] |= HEX;
    for (int i='A'; i<='F'; ++i) charMap[i] |= HEX;
  }

//////////////////////////////////////////////////////////////////////////
// Locale
//////////////////////////////////////////////////////////////////////////

  public static boolean localeIsUpper(long self)
  {
    return Character.isUpperCase((int)self);
  }

  public static boolean localeIsLower(long self)
  {
    return Character.isLowerCase((int)self);
  }

  public static long localeUpper(long self)
  {
    // Java doesn't provide a locale Character API
    return Character.toString((char)self).toUpperCase(Locale.current().java()).charAt(0);
  }

  public static long localeLower(long self)
  {
    // Java doesn't provide a locale Character API
    return Character.toString((char)self).toLowerCase(Locale.current().java()).charAt(0);
  }

//////////////////////////////////////////////////////////////////////////
// Conversion
//////////////////////////////////////////////////////////////////////////

  public static String toChar(long self)
  {
    if (self < 0 || self > 0xFFFF) throw Err.make("Invalid unicode char: " + self).val;
    if (self < FanStr.ascii.length) return FanStr.ascii[(int)self];
    return String.valueOf((char)self);
  }

  public static String toHex(long self) { return toHex(self, null); }
  public static String toHex(long self, Long width)
  {
    String s = Long.toHexString(self);
    if (width != null && s.length() < width.intValue())
      s = zeros[width.intValue()-s.length()] + s;
    return s;
  }
  static String[] zeros = new String[16];
  static { zeros[0] = ""; for (int i=1; i<zeros.length; ++i) zeros[i] = zeros[i-1] + "0"; }

  public static String toStr(long self)
  {
    return String.valueOf(self);
  }

  public static String toCode(long self) { return String.valueOf(self); }
  public static String toCode(long self, long base)
  {
    if (base == 10) return String.valueOf(self);
    if (base == 16) return "0x" + Long.toHexString(self);
    throw ArgErr.make("Invalid base " + base).val;
  }

//////////////////////////////////////////////////////////////////////////
// Closures
//////////////////////////////////////////////////////////////////////////

  public static void times(long self, Func f)
  {
    for (long i=0; i<self; ++i)
      f.call1(Long.valueOf(i));
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  /** sys::Int.maxValue */
  public static final long maxValue = Long.MAX_VALUE;

  /** sys::Int.minValue */
  public static final long minValue = Long.MIN_VALUE;

  /** sys::Int.defVal */
  public static final long defVal = 0L;

  // default chunk size for IO buffering (defaults to 4KB)
  public static final Long Chunk = Long.valueOf(4096);

  // internalized boxed Longs used for byte IO
  static final Long[] pos = new Long[256];
  static { for (int i=0; i<pos.length; ++i) pos[i] = Long.valueOf(i); }

}

File: src/nfan/fan/sys/FanInt.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   13 Sep 06  Andy Frank  Creation
//

using System;
using Fanx.Serial;

namespace Fan.Sys
{
  /// <summary>
  /// FanInt defines the methods for sys::Int.  The actual
  /// class used for representation is Fan.Sys.Int.
  /// </summary>
  public sealed class FanInt
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Long fromStr(string s) { return fromStr(s, 10, true); }
    public static Long fromStr(string s, long radix) { return fromStr(s, radix, true); }
    public static Long fromStr(string s, long radix, bool check)
    {
      try
      {
        return Long.valueOf(Convert.ToInt64(s, (int)radix));
      }
      catch (OverflowException)
      {
        if (!check) return null;
        throw ParseErr.make("Int", s).val;
      }
      catch (FormatException)
      {
        if (!check) return null;
        throw ParseErr.make("Int", s).val;
      }
    }

    public static long random() { return random(null); }
    public static long random(Range r)
    {
      rand.GetBytes(randBytes);
      long v = BitConverter.ToInt64(randBytes, 0);
      if (r == null) return v;
      if (v < 0) v = -v;
      long start = r.start();
      long end   = r.end();
      if (r.inclusive()) ++end;
      return start + (v % (end-start));
    }

    static byte[] randBytes = new byte[8];
    static readonly System.Security.Cryptography.RNGCryptoServiceProvider
      rand = new System.Security.Cryptography.RNGCryptoServiceProvider();

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public static bool equals(long self, object obj)
    {
      if (obj is Long)
        return self == (obj as Long).longValue();
      else
        return false;
    }

    public static long compare(long self, object obj)
    {
      long val = self;
      long that = ((long)obj);
      if (val < that) return -1; return val == that ? 0 : +1;
    }

    public static long hash(long self)
    {
      return self;
    }

    public static  Type type(long self)
    {
      return Sys.IntType;
    }

  //////////////////////////////////////////////////////////////////////////
  // Operators
  //////////////////////////////////////////////////////////////////////////

    public static long negate    (long self)         { return -self; }
    public static long inverse   (long self)         { return ~self; }
    public static long mult      (long self, long x) { return self * x; }
    public static long div       (long self, long x) { return self / x; }
    public static long mod       (long self, long x) { return self % x; }
    public static long plus      (long self, long x) { return self + x; }
    public static long minus     (long self, long x) { return self - x; }
    public static long and       (long self, long x) { return self & x; }
    public static long or        (long self, long x) { return self | x; }
    public static long xor       (long self, long x) { return self ^ x; }
    public static long lshift    (long self, long x) { return self << ((int)x); }
    public static long rshift    (long self, long x) { return self >> ((int)x); }
    public static long increment (long self)         { return self+1; }
    public static long decrement (long self)         { return self-1; }

  //////////////////////////////////////////////////////////////////////////
  // Math
  //////////////////////////////////////////////////////////////////////////

    public static long abs(long self)
    {
      long val = self;
      if (val >= 0) return self;
      return -val;
    }

    public static long min(long self, long that)
    {
      long val = self;
      if (val <= that) return self;
      return that;
    }

    public static long max(long self, long that)
    {
      long val = self;
      if (val >= that) return self;
      return that;
    }

    public static bool isEven(long self)
    {
      return (self % 2) == 0;
    }

    public static bool isOdd(long self)
    {
      return (self % 2) != 0;
    }

  /////////////////////////////////////////////////////////////////////////
  // Char
  //////////////////////////////////////////////////////////////////////////

    public static bool isSpace(long self)
    {
      try
      {
        return (self < 128 && (charMap[(int)self] & SPACE) != 0);
      }
      catch (IndexOutOfRangeException)
      {
        // should be very rare to use this method with negative
        // numbers, so don't take the hit every call
        return false;
      }
    }

    public static bool isAlpha(long self)
    {
      try
      {
        return (self < 128 && (charMap[(int)self] & ALPHA) != 0);
      }
      catch (IndexOutOfRangeException)
      {
        // should be very rare to use this method with negative
        // numbers, so don't take the hit every call
        return false;
      }
    }

    public static bool isAlphaNum(int self)
    {
      try
      {
        return (self < 128 && (charMap[(int)self] & ALPHANUM) != 0);
      }
      catch (IndexOutOfRangeException)
      {
        // should be very rare to use this method with negative
        // numbers, so don't take the hit every call
        return false;
      }
    }

    public static bool isUpper(long self)
    {
      long val = self;
      return 'A' <= val && val <= 'Z';
    }

    public static bool isLower(long self)
    {
      long val = self;
      return 'a' <= val && val <= 'z';
    }

    public static long upper(long self)
    {
      long val = self;
      if ('a' <= val && val <= 'z')
        return ((int)val) & ~0x20;
      else
        return self;
    }

    public static long lower(long self)
    {
      long val = self;
      if ('A' <= val && val <= 'Z')
        return ((int)val) | 0x20;
      else
        return self;
    }

    public static bool isDigit(long self)
    {
      long val = self;
      return '0' <= val && val <= '9';
    }

    public static bool isDigit(long self, long radix)
    {
      int val = (int)self;
      if (val < 0 || val >= 128) return false;
      int r   = (int)radix;

      if (r == 10)
      {
        return ((charMap[val] & DIGIT) != 0);
      }

      if (r == 16)
      {
        return ((charMap[val] & HEX) != 0);
      }

      if (r <= 10)
      {
        return '0' <= val && val <= ('0'+r);
      }
      else
      {
        if ((charMap[val] & DIGIT) != 0) return true;
        int x = val-10;
        if ('a' <= val && val <= 'a'+x) return true;
        if ('A' <= val && val <= 'A'+x) return true;
        return false;
      }
    }

    public static Long toDigit(long self)
    {
      long val = self;
      if (0 <= val && val <= 9) return Long.valueOf((int)val + '0');
      return null;
    }

    public static Long toDigit(long self, long radix)
    {
      int val = (int)self;
      int r   = (int)radix;
      if (val < 0 || val >= r) return null;

      if (val < 10) return Long.valueOf(val + '0');
      return Long.valueOf(val - 10 + 'a');
    }

    public static Long fromDigit(long self)
    {
      long val = self;
      if ('0' <= val && val <= '9') return Long.valueOf((int)val - '0');
      return null;
    }

    public static Long fromDigit(long self, long radix)
    {
      int val = (int)self;
      if (val < 0 || val >= 128) return null;
      int r   = (int)radix;

      int ten = r < 10 ? r : 10;
      if ('0' <= val && val < '0'+ten) return Long.valueOf(val - '0');
      if (r > 10)
      {
        int alpha = r-10;
        if ('a' <= val && val < 'a'+alpha) return Long.valueOf(val + 10 - 'a');
        if ('A' <= val && val < 'A'+alpha) return Long.valueOf(val + 10 - 'A');
      }
      return null;
    }

    public static bool equalsIgnoreCase(long self, long ch)
    {
      long val = self;
      return (val | 0x20L) == (ch | 0x20L);
    }

    internal static readonly byte[] charMap = new byte[128];
    internal static readonly byte SPACE   = 0x01;
    internal static readonly byte UPPER   = 0x02;
    internal static readonly byte LOWER   = 0x04;
    internal static readonly byte DIGIT   = 0x08;
    internal static readonly byte HEX     = 0x10;
    internal static readonly int ALPHA    = UPPER | LOWER;
    internal static readonly int ALPHANUM = UPPER | LOWER | DIGIT;

    // static initializer below

  //////////////////////////////////////////////////////////////////////////
  // Locale
  //////////////////////////////////////////////////////////////////////////

    public static bool localeIsUpper(long self)
    {
      long val = self;
      return Char.IsUpper((char)val);
    }

    public static bool localeIsLower(long self)
    {
      long val = self;
      return Char.IsLower((char)val);
    }

    public static long localeUpper(long self)
    {
      long val = self;
      return Char.ToUpper((char)val, Locale.current().dotnet());
    }

    public static long localeLower(long self)
    {
      long val = self;
      return Char.ToLower((char)val, Locale.current().dotnet());
    }

  //////////////////////////////////////////////////////////////////////////
  // Conversion
  //////////////////////////////////////////////////////////////////////////

    public static string toChar(long self)
    {
      long val = self;
      if (val < 0 || val > 0xFFFF) throw Err.make("Invalid unicode char: " + val).val;
      if (val < FanStr.m_ascii.Length) return FanStr.m_ascii[(int)val];
      return "" + (char)val;
    }

    public static string toHex(long self) { return toHex(self, null); }
    public static string toHex(long self, Long width)
    {
      long val = self;
      string s = val.ToString("X").ToLower();
      if (width != null && s.Length < width.intValue())
        s = zeros[width.intValue()-s.Length] + s;
      return s;
    }
    static string[] zeros = new string[16];

    public static string toStr(long self)
    {
      return self.ToString();
    }

    public static void encode(long self, ObjEncoder @out)
    {
      @out.w(self.ToString());
    }

  //////////////////////////////////////////////////////////////////////////
  // Closures
  //////////////////////////////////////////////////////////////////////////

    public static void times(long self, Func f)
    {
      long val = self;
      for (long i=0; i<val; i++)
        f.call1(i);
    }

  //////////////////////////////////////////////////////////////////////////
  // Intern
  //////////////////////////////////////////////////////////////////////////

    /// <summary>sys::Int.maxValue</summary>
    public static readonly long m_maxValue = Int64.MaxValue;

    /// <summary>sys::Int.maxValue</summary>
    public static readonly long m_minValue = Int64.MinValue;

    /// <summary>sys::Int.defVal</summary>
    public static readonly long m_defVal = 0;

    // default check size for IO buffering (defaults to 4KB)
    public static readonly Long Chunk = Long.valueOf(4096);

    static FanInt()
    {
      charMap[' ']  |= SPACE;
      charMap['\n'] |= SPACE;
      charMap['\r'] |= SPACE;
      charMap['\t'] |= SPACE;
      charMap['\f'] |= SPACE;

      // alpha characters
      for (int i='a'; i<='z'; i++) charMap[i] |= LOWER;
      for (int i='A'; i<='Z'; i++) charMap[i] |= UPPER;

      // digit characters
      for (int i='0'; i<='9'; i++) charMap[i] |= DIGIT;

      // hex characters
      for (int i='0'; i<='9'; i++) charMap[i] |= HEX;
      for (int i='a'; i<='f'; i++) charMap[i] |= HEX;
      for (int i='A'; i<='F'; i++) charMap[i] |= HEX;

      // Zeros
      zeros[0] = "";
      for (int i=1; i<zeros.Length; ++i)
        zeros[i] = zeros[i-1] + "0";
    }

  }
}